server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveClusterAction.java

```diff
@@ -0,0 +1,455 @@
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ResolveClusterAction.Response response = (ResolveClusterAction.Response) o;
+            return infoMap.equals(response.infoMap);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(infoMap);
+        }
+
+        public Map<String, ResolveClusterInfo> getResolveClusterInfo() {
+            return infoMap;
+        }
+    }
+
+    public static class TransportAction extends HandledTransportAction<ResolveClusterAction.Request, ResolveClusterAction.Response> {
+
+        private final ThreadPool threadPool;
+        private final ClusterService clusterService;
+        private final RemoteClusterService remoteClusterService;
+        private final IndexNameExpressionResolver indexNameExpressionResolver;
+        private final boolean ccsCheckCompatibility;
+
+        @Inject
+        public TransportAction(
+            TransportService transportService,
+            ClusterService clusterService,
+            ThreadPool threadPool,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver
+        ) {
+            super(NAME, transportService, actionFilters, ResolveClusterAction.Request::new, EsExecutors.DIRECT_EXECUTOR_SERVICE);
+            this.threadPool = threadPool;
+            this.clusterService = clusterService;
+            this.remoteClusterService = transportService.getRemoteClusterService();
+            this.indexNameExpressionResolver = indexNameExpressionResolver;
+            this.ccsCheckCompatibility = SearchService.CCS_VERSION_CHECK_SETTING.get(clusterService.getSettings());
+        }
+
+        @Override
+        protected void doExecute(Task task, ResolveClusterAction.Request request, ActionListener<ResolveClusterAction.Response> listener) {
+            if (ccsCheckCompatibility) {
+                checkCCSVersionCompatibility(request);
+            }
+            ClusterState clusterState = clusterService.state();
+            Map<String, OriginalIndices> remoteClusterIndices = remoteClusterService.groupIndices(
+                request.indicesOptions(),
+                request.indices()
+            );
+            OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+
+            Map<String, ResolveClusterInfo> clusterInfoMap = new ConcurrentHashMap<>();
+            if (remoteClusterIndices.size() > 0) {
+                int remoteRequests = remoteClusterIndices.size();
+                CountDown completionCounter = new CountDown(remoteRequests);
+                Runnable terminalHandler = () -> {
+                    if (completionCounter.countDown()) {
```

Comment:
I think you could replace this with `RefCountingRunnable` which would let you simplify some of this logic a bit (it handles the zero case properly)

Reply:
Fixed in [this commit](https://github.com/elastic/elasticsearch/pull/102726/commits/d1f6d81226d3416685abaae358a936369963675d).